cpp/src/arrow/dataset/scanner.h

```diff
@@ -29,26 +29,61 @@
 namespace arrow {
 namespace dataset {
 
-/// \brief Shared state for a Scan operation
-struct ARROW_DS_EXPORT ScanContext {
-  MemoryPool* pool = arrow::default_memory_pool();
-};
+/// \brief Base class for file scanning options
+class ARROW_DS_EXPORT FileScanOptions {
+ public:
+  /// \brief The file format this options corresponds to
+  virtual std::shared_ptr<FileFormat> file_format() const = 0;
 
-// TODO(wesm): API for handling of post-materialization filters. For
-// example, if the user requests [$col1 > 0, $col2 > 0] and $col1 is a
-// partition key, but $col2 is not, then the filter "$col2 > 0" must
-// be evaluated in-memory against the RecordBatch objects resulting
-// from the Scan
+  virtual ~FileScanOptions() = default;
+};
 
-class ARROW_DS_EXPORT ScanOptions {
+struct ARROW_DS_EXPORT ScanContext final {
```

Comment:
I think this should still be named `ScanOptions`. Perhaps you can use `ScanContext` for some opaque internal state (if necessary).

Reply:
Why should it be named ScanOptions? It isn't optional, unlike the potentially absent FileScanOptions. Furthermore that name implies that FileScanOptions inherits ScanOptions, which is no longer the case. If ScanContext isn't a good name then what about ScanState?

Reply:
Because those are options, simply.

Reply:
But we also have `compute::FunctionContext` which is required by the kernels, and may optionally be configured with a different memory pool

Reply:
Lots of things can be configured with a memory pool. But `FunctionContext` doesn't hold any other options; it's a container for kernel state execution, not user-provided preferences.

Reply:
Both argument are valid, it holds a MemoryPool like FunctionContext, but it has options like other kernel. I'm biased toward ScanOptions as it holds important options like the projected schema and the filters.

Reply:
Alright, I'll rename back to ScanOptions.